Add intermediary commit step to storage transactions.

dandanlen opened the pull request:
Addresses issue #102 - I came across the issue on Workshub and decided to give it a go. Any feedback welcome.

ethanfrey commented on src/storage.rs:
```diff
@@ -2,42 +2,62 @@ use crate::errors::Result;
-impl<'a, S: Storage> StorageTransaction<'a, S> {
-    pub fn new(storage: &'a mut S) -> Self {
+impl Op {
+    pub fn apply<S: Storage>(&self, storage: &mut S) {
```
Nice to pull this into a method on `Op` rather than body of `commit`

ethanfrey commented on src/storage.rs:
```diff
@@ -95,11 +115,30 @@ mod test {
         assert_eq!(check.get(b"foo"), Some(b"bar".to_vec()));
         check.set(b"subtx", b"works");
-        check.commit();
+        check.prepare().commit(&mut base);
```
Nice, the syntax did work as hoped

ethanfrey commented on src/storage.rs:
```diff
@@ -95,11 +115,30 @@ mod test {
+    }
+
+    #[test]
+    fn storage_remains_readable() {
```
Good test case

dandanlen commented:
Hi @ethanfrey, thanks for the feedback, glad I could be of help. While adding some rustdoc (in the same minimalist style as the existing docstrings 😉 ) I decided on a minor reorg... 

I renamed `Commit` to `RepLog` and added it to the `StorageTransaction` struct instead of a naked vec. This prevents passing the Vec around - in the `prepare` method we just have to consume `self` and return the RefLog. 

I think this makes things a little clearer - if you disagree I would happy to roll back to the previous code, just let me know.

ethanfrey commented:
Nice one, I like it.

ethanfrey commented on src/storage.rs:
```diff
@@ -68,11 +77,12 @@ impl<'a, S: ReadonlyStorage> ReadonlyStorage for StorageTransaction<'a, S> {
             value: value.to_vec(),
-        })
+        };
+        op.apply(&mut self.local_state);
```
I like this, use same `apply` both for local cache and later when committing the `rep_log`

ethanfrey commented:
Thanks again, and if you don't mind I will ping you in my next PR that makes use of this https://github.com/CosmWasm/cosmwasm/issues/90 but it's really needed to let us query state in a tx.

dandanlen commented:
👍 yup, no problem.
